File: go/internal/sol/p0062/p0062.go
package p0062

import (
	"fmt"
	"slices"

	"pe-solver/internal/mylib"
)

func compute(size int) string {
	key := func(n int) string {
		tmp := []rune(fmt.Sprintf("%d", n))
		slices.Sort(tmp)

		return string(tmp)
	}

	tbl := map[string][]int{}
	for n := 1; ; n++ {
		id := key(n * n * n)
		if v, ok := tbl[id]; ok {
			tbl[id] = append(v, n)
			if len(tbl[id]) == size {
				return fmt.Sprintf("%d", mylib.Pow(tbl[id][0], 3))
			}
		} else {
			tbl[id] = []int{n}
		}
	}
}

func Solve() string {
	return compute(5)
}

File: haskell/src/Sol/P0062.hs
module Sol.P0062 (compute, solve) where

import Data.List (sort)

import qualified Data.Map.Strict as M (Map, empty, insert, lookup)

import Mylib.Util (lastExn)

compute :: Int -> String
compute len =
    show $ aux M.empty [1 ..]
  where
    aux :: M.Map String [Int] -> [Int] -> Int
    aux _ [] = error "unreachable"
    aux tbl (n : ns)
        | Just v <- M.lookup key tbl =
            if length v == len - 1
                then lastExn v
                else aux (M.insert key (cube : v) tbl) ns
        | otherwise =
            aux (M.insert key [cube] tbl) ns
      where
        cube = n ^ (3 :: Int)
        key = sort $ show cube

solve :: String
solve = compute 5
